runtime/runtime_events.c

```diff
@@ -404,18 +405,17 @@ static void runtime_events_create_raw(void) {
   }
 }
 
-/* Stop the world section which calls [runtime_events_create_raw], used when we
-   can't be sure there is only a single domain running. */
-static void
-stw_create_runtime_events(caml_domain_state *domain_state, void *data,
-                              int num_participating,
-                              caml_domain_state **participating_domains) {
-  /* Everyone must be stopped for starting and stopping runtime_events */
+static void stw_create_runtime_events(
+  caml_domain_state *domain_state, void *data,
+  int num_participating,
+  caml_domain_state **participating_domains)
+{
+  /* TODO: this global_barrier is not necessary. */
```

Comment:
Just to check I get this right: it is not required because it is called from a STW segment and the barrier is therefore not needed? (since it runs from a single domain that is already checked for in the next block?)

Reply:
My reasoning is the following.

STW callbacks only start running when all registered domains have synchronized, which amounts to a barrier -- you can see barrier-looking code in at the beginning of `stw_handler` in domain.c; for the leader the barrier behavior comes from the calls to `wait_interrupt_serviced`.

Additional/explicit barriers are only useful in the middle or at the end of a STW callback, to guarantee that the callback logic before the barrier does not race with the callback logic after the barrier or (at the end) with mutators resuming after the STW callback.

Reply:
I don't believe this is correct. The STW code uses the runtime events API between the "enter barrier" and the time the domain starts running the STW section [here](https://github.com/ocaml/ocaml/blob/a912cff51de76b64c8341e39e4ec60361724553b/runtime/domain.c#L1345). That is, it's not mutator code we'd be "racing" with, but the runtime there. Removing this barrier coincidentally wouldn't be a data race, I believe, but it would mean that orphan `END` events could be emitted on the line I linked, depending on how the threads happen to be scheduled. Removing the other similar double barrier in the teardown STW does cause a race and horrible things.

It should be sufficient to have one single barrier use, running the setup/teardown from the final party while all others are blocked waiting for the release. In #12579 I do this and add a comment, similar to my explanation above, detailing why it has to be done this way.

Reply:
Thanks for your comments, I had indeed not considered the case of races between the code here and the instruction inside the STW runner functions. I will remove the offending TODOs and leave any change there to your PR. (In particular, I am happy to delegate to you the responsibility of document the subtle thing that is going on there.)